### PR TITLE
feat(ddm): scratchpads

### LINKED ISSUE #57269: Ability to save the exploration
- have a way to save the exploration in local storage with a simple recall dropdown

### PR DESCRIPTION
**Pull request by obostjancic**
Closes https://github.com/getsentry/sentry/issues/57269

Adds ability to save current selection of widgets as scratchpad in local storage. Adds basic Scratchpad management.

https://github.com/getsentry/sentry/assets/86684834/358c8018-39dc-4e07-a11a-501d4946831c

### CODE REVIEW

File: static/app/views/ddm/metricScratchpad.tsx [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+  // changes the query when a scratchpad is selected, clears it when none is selected
+  useEffect(() => {
+    if (scratchpads.selected) {
+      const selectedQuery = scratchpads.all[scratchpads.selected].query;
+      updateQuery(router, selectedQuery);
+    } else if (scratchpads.selected === null) {
+      clearQuery(router);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [scratchpads.selected]);
```
Comment (ArthurKnaus) [inline static/app/views/ddm/metricScratchpad.tsx:129]: I wonder whether the whole router syncing should be in `useScratchpad` 🤔 
The first useEffect could then be replaced by a simple default value for selected, right?

File: static/app/views/ddm/metricScratchpad.tsx [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+                      message: t('Are you sure you want to delete this scratchpad?'),
+
+                      confirmText: t('Delete'),
+                    });
+                  }}
+                />
+              </StyledDropdownIcon>
+            </Fragment>
+          ),
+        }))}
```
Comment (ArthurKnaus) [inline static/app/views/ddm/metricScratchpad.tsx:175]: I would `useMemo` on the options.

File: static/app/views/ddm/metricScratchpad.tsx [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        value={scratchpads.selected ?? ''}
+        closeOnSelect={false}
+        onChange={option => {
+          if (option.value === scratchpads.selected) {
+            scratchpads.select(null);
+          } else {
+            scratchpads.select(option.value);
+          }
+        }}
```
Comment (ArthurKnaus) [inline static/app/views/ddm/metricScratchpad.tsx:184]: This duplicates the setting default logic. You could have one `toggleDefault` function for that.
```ts
function toggleDefault(value?: string) {
  if (value === scratchpads.selected) {
    scratchpads.select(null);
  } else {
    scratchpads.select(value ?? null);
  }
} 
```

**Review by ArthurKnaus — APPROVED** [review]
nice work 🚀

File: static/app/views/ddm/metricScratchpad.tsx [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            <Fragment>
+              <StyledDropdownIcon>
+                <IconDelete
+                  onClick={() => {
+                    openConfirmModal({
+                      onConfirm: () => scratchpads.remove(s.id),
+                      message: t('Are you sure you want to delete this scratchpad?'),
+
+                      confirmText: t('Delete'),
+                    });
```
Comment (ArthurKnaus) [inline static/app/views/ddm/metricScratchpad.tsx:169]:
```suggestion
                    openConfirmModal({
                      onConfirm: () => scratchpads.remove(s.id),
                      message: t('Are you sure you want to delete this scratchpad?'),
                      confirmText: t('Delete'),
                    });
```

File: static/app/views/ddm/metricScratchpad.tsx (hunk 1 above)
Comment (obostjancic) [reply static/app/views/ddm/metricScratchpad.tsx]: Yep, much nicer!

**Review by matejminar — APPROVED** [review]
Looks good to me. Two pieces of feedback:
1. Should these use the same size?
image
2. Can we make this submittable with Enter? And can we clear the input and close it after Enter / clicking on save?
image

**Pull request merged**